File: articles/nonlinearity-feedback-biological-regulation/fortran/nonlinear_feedback_kernel.f90
! Compact nonlinear feedback numerical kernel in Fortran.

program nonlinear_feedback_kernel
  implicit none

  integer :: step, steps
  real :: response_sat, response_hill
  real :: signal, vmax, k_half, hill_n
  real :: x, set_point, k, dx, dt
  real :: alpha, beta, production, loss

  signal = 20.0
  vmax = 1.0
  k_half = 20.0
  response_sat = vmax * signal / (k_half + signal)

  signal = 60.0
  k_half = 40.0
  hill_n = 4.0
  response_hill = signal**hill_n / (k_half**hill_n + signal**hill_n)

  dt = 0.05
  x = 180.0
  set_point = 100.0
  k = 0.18
  steps = int(30.0 / dt) + 1

  do step = 2, steps
    dx = -k * (x - set_point)
    x = x + dx * dt
  end do

  print *, "Saturating response at 20:", response_sat
  print *, "Hill response at 60:", response_hill
  print *, "Negative feedback final:", x

  dt = 0.01
  x = 2.0
  alpha = 3.0
  beta = 0.8
  k_half = 1.5
  hill_n = 4.0
  steps = int(80.0 / dt) + 1

  do step = 2, steps
    production = alpha * x**hill_n / (k_half**hill_n + x**hill_n)
    loss = beta * x
    dx = production - loss
    x = max(x + dx * dt, 0.0)
  end do

  print *, "Positive feedback final:", x
end program nonlinear_feedback_kernel

File: articles/nonlinearity-feedback-biological-regulation/c/nonlinear_feedback_kernel.c
/*
 * Compact nonlinear feedback numerical kernel in C.
 */

#include <math.h>
#include <stdio.h>

double saturating_response(double signal, double vmax, double k_half) {
    return vmax * signal / (k_half + signal);
}

double hill_response(double signal, double k_half, double n) {
    return pow(signal, n) / (pow(k_half, n) + pow(signal, n));
}

double negative_feedback_final(double x0, double set_point, double k, double dt, double t_end) {
    int steps = (int)floor(t_end / dt) + 1;
    double x = x0;

    for (int i = 1; i < steps; i++) {
        double dx = -k * (x - set_point);
        x += dx * dt;
    }

    return x;
}

double positive_feedback_final(double x0, double alpha, double beta, double k_half, double n, double dt, double t_end) {
    int steps = (int)floor(t_end / dt) + 1;
    double x = x0;

    for (int i = 1; i < steps; i++) {
        double production = alpha * pow(x, n) / (pow(k_half, n) + pow(x, n));
        double loss = beta * x;
        double dx = production - loss;
        x = fmax(x + dx * dt, 0.0);
    }

    return x;
}

int main(void) {
    printf("saturating_at_20=%.6f\n", saturating_response(20.0, 1.0, 20.0));
    printf("hill_at_60_n4=%.6f\n", hill_response(60.0, 40.0, 4.0));
    printf("negative_feedback_final=%.6f\n", negative_feedback_final(180.0, 100.0, 0.18, 0.05, 30.0));
    printf("positive_feedback_final=%.6f\n", positive_feedback_final(2.0, 3.0, 0.8, 1.5, 4.0, 0.01, 80.0));
    return 0;
}
